src/libmongoc/src/mongoc/mongoc-uri.c

```diff
@@ -1526,6 +1526,23 @@ mongoc_uri_finalize_directconnection (mongoc_uri_t *uri, bson_error_t *error)
    return true;
 }
 
+static void
+mongoc_uri_check_for_genuine_hosts (const char *hosts)
+{
+   char *hosts_lowercase = lowercase_str_new (hosts);
```

Comment:
In https://github.com/mongodb-js/mongodb-build-info/commit/52f4b0d73eaeeecd85598c07e89033fb3f5b1f16, the shell uses an anchored regex with case-insensitive match on _each_ host. I opted to keep things simple by searching the entire host string, but kept the case-insensitive behavior.

I briefly looked into [`strcasestr`](https://linux.die.net/man/3/strcasestr) but found not prior art in libmongoc for that. It looks like it may be non-portable, and I've had my fill of that after the `strerror_r()` debacle.

There a notably no tests for a case-insensitive match, but I'm happy to add some if requested. The PR currently uses the same fixtures from the shell (where case-insensitive matching also seems to be untested).

Reply:
Thanks for pointing out `ends_with`. I missed that previously.

It wasn't clear from [DRIVERS-2583](https://jira.mongodb.org/browse/DRIVERS-2583) whether we need to log messages for _each_ host or just once. I'm inclined to say that once is enough to avoid log pollution -- and also because it seems unlikely you'd ever have a mixed cluster (i.e. all hosts will either be genuine or not). And a single string match across the entire host/port component of the URI seemed OK to limit overhead and ensure at most one log message. As for matching within the string vs. using `ends_with`, I also think it's unlikely that one of these substrings would appear in a genuine host (e.g. `mydb.cosmos.azure.com.mydomain.com`) so I wasn't worried about that.

Furthermore, I think it's reasonable to just check the initial SRV host string, knowing that a Cosmos connection string is going to resolve to more Cosmos hosts. If someone is going out of their way to configure SRV with their own domain and point it to a Cosmos or DocumentDB, I'd argue they're probably not the intended audience for this log message.

Reply:
Resolving this, as I'm now using `mongoc_ends_with()` to do a rear-anchored match.